compiler/rustc_builtin_macros/src/edition_panic.rs

```diff
@@ -19,50 +40,99 @@ pub(crate) fn expand_panic<'cx>(
     sp: Span,
     tts: TokenStream,
 ) -> MacroExpanderResult<'cx> {
-    let mac = if use_panic_2021(sp) { sym::panic_2021 } else { sym::panic_2015 };
+    let mac = if use_panic_2021(sp) { InnerCall::Panic2021 } else { InnerCall::Panic2015 };
     expand(mac, cx, sp, tts)
 }
 
 /// This expands to either
 /// - `$crate::panic::unreachable_2015!(...)` or
 /// - `$crate::panic::unreachable_2021!(...)`
-/// depending on the edition.
+/// depending on the edition. If the entire message is known at compile time,
+/// `core::panicking::panic` may be called as an optimization.
 pub(crate) fn expand_unreachable<'cx>(
     cx: &'cx mut ExtCtxt<'_>,
     sp: Span,
     tts: TokenStream,
 ) -> MacroExpanderResult<'cx> {
-    let mac = if use_panic_2021(sp) { sym::unreachable_2021 } else { sym::unreachable_2015 };
+    let mac =
+        if use_panic_2021(sp) { InnerCall::Unreachable2021 } else { InnerCall::Unreachable2015 };
     expand(mac, cx, sp, tts)
 }
 
 fn expand<'cx>(
-    mac: rustc_span::Symbol,
+    mac: InnerCall,
     cx: &'cx ExtCtxt<'_>,
     sp: Span,
     tts: TokenStream,
 ) -> MacroExpanderResult<'cx> {
     let sp = cx.with_call_site_ctxt(sp);
 
+    // If the call is of the form `panic!(<string literal>)` and there are no formatting arguments
+    // in the string literal, we can call `core::panicking::panic` to centralize the panic logic.
+    if tts.len() == 1
+        && let Some(TokenTree::Token(token, _)) = tts.get(0)
+        && let TokenKind::Literal(lit) = &token.kind
+        && let Lit { kind: LitKind::Str | LitKind::StrRaw(_), symbol, .. } = lit
+        && let msg = symbol.as_str()
+        && !msg.contains(|c| c == '{' || c == '}')
```

Comment:
This is very much a hack. It would be much better to use `rustc_parse_format` here.

In particular since I think your version would not optimize `panic!("foo {{ bar }}");` since it doesn't know about the escaping.